src/hotspot/share/opto/predicates.cpp

```diff
@@ -147,3 +149,170 @@ Node* PredicateBlock::skip_regular_predicates(Node* regular_predicate_proj, Deop
+  // of diamond with it:
+  //
+  //        C3       X
+  //       /  \     /
+  //     C2   current
+  //      \  /
+  //       C1
+  //
+  // Cx means collected. Since C3 is already collected (and thus already visited), we add 'current' to the collected list
+  // since it must also be on a path from start->target. We continue the DFS with X which could potentially also be on a
+  // start->target path but that is not known yet.
+  //
+  // This method returns true if there is an unvisited input and return false otherwise if all inputs have been visited.
+  bool push_next_unvisited_input() {
+    Node* current = _stack.node();
+    const uint next_unvisited_input_index = _stack.index();
+    for (uint i = next_unvisited_input_index; i < current->req(); i++) {
+      Node* input = current->in(i);
+      if (_node_filter(input)) {
+        if (!_visited.test(input->_idx)) { // Avoid double visits which could take a long time to process.
+          // Visit current->in(i) next in DFS walk. Once 'current' is again on top of stack, we need to visit in(i+1).
+          push_input_and_update_current_index(input, i + 1);
+          return true;
+        } else if (_collected_nodes.member(input)) {
+          // Diamond case, see description above.
+          // Input node part of start->target? Then current node (i.e. a predecessor of input) is also on path. Collect it.
+          _collected_nodes.push(current);
+        }
+      }
+    }
+    return false;
+  }
+
+  // Update the index of the current node on top of the stack with the next unvisited input index and push 'input' to
+  // the stack which is visited next in the DFS order.
+  void push_input_and_update_current_index(Node* input, uint next_unvisited_input_index) {
+    _stack.set_index(next_unvisited_input_index);
+    push_unvisited_node(input);
+  }
+
+  // Push the next unvisited node in the DFS order with index 1 since this node needs to visit all its inputs.
+  void push_unvisited_node(Node* next_to_visit) {
+    _stack.push(next_to_visit, 1);
+  }
+
+  // If the current node on top of the stack is on a path from start->target(s), then also collect the predecessor node
+  // before popping the current node.
+  void pop_node_and_maybe_collect_predecessor() {
+    Node* current_node = _stack.node();
+    _stack.pop();
+    if (_stack.is_nonempty() && _collected_nodes.member(current_node)) {
+      // Current node was part of start->target? Then predecessor (i.e. newly on top of stack) is also on path. Collect it.
+      Node* predecessor = _stack.node();
+      _collected_nodes.push(predecessor);
+    }
+  }
+};
+
+// Clones this Template Assertion Predicate Expression and applies the given strategy to transform the OpaqueLoop* nodes.
+Opaque4Node* TemplateAssertionPredicateExpression::clone(TransformStrategyForOpaqueLoopNodes& transform_strategy,
+                                                         Node* new_ctrl, PhaseIdealLoop* phase) {
+  ResourceMark rm;
```

Comment:
The ResourceMark makes me a bit nervous, in combination of a non-constant `transform_strategy`.
Could the `transform_strategy` be a constant reference, maybe by making its functions also const?